docs: add over-engineering post (#3)
43081j/43081j.github.io: _posts/2025-09-09-overengineering-inputs.md

---
layout: post

> title: Over-engineering inputs
> description: Thoughts on how the ecosystem has suffered from over-engineering of input validation and edge case

Reviewer: I like how these sounds already, but for me the title doesn't make it clear what this article is about. Sounds a bit too generic (?). For me `edge-case-first` sounds like a good catch phrase (and you use it already inside the article)

```suggestion
title: Edge‑case‑first APIs are why your deps are bloated
description: How chasing edge cases led to atomized packages, surprise behavior, and heavier dependency trees across the JS ecosystem
```

---
layout: post
title: Over-engineering inputs
description: Thoughts on how the ecosystem has suffered from over-engineering of input validation and edge case
---

This is just some of what I've been pondering recently - particularly in terms of how we ended up with such overly-granular dependency trees.

> I think we've ended up with many libraries in our ecosystem which are edge-case-first, the opposite to what I'd expect. I'll give a few examples and some thoughts around this, mostly in the hope we can start to trim some of it away.

Reviewer: ```suggestion
We’ve normalized edge-case-first libraries - the opposite of what I’d expect. I'll give a few examples and some thoughts around this, mostly in the hope we can start to trim some of it away.
```

"normalized" may be not the best choice of words from me. But I hope you get the idea

# The problem

I believe a lot of the questionably small libraries hiding in our deep dependency trees are a result of over-engineering for inputs and edge cases we've probably never seen.

> For example, let's say we want to build a `clamp` library:

Reviewer: ```suggestion
For example, say we’re building a simple clamp:
```

Some folks are shipping apps/features, not libraries. Soo.. This might be better

```ts
export function clamp(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}
```

Pretty simple!

> Though, what if someone passes nonsensical ranges? Let's add some validation:

Reviewer: ```suggestion
But what if someone passes nonsensical ranges? Let's add a tiny check:
```

isRegex(iframeRegex); // true
```

This is indeed useful, and I do support this myself in chai (which I maintain). However, this is an edge case most libraries don't need to care about.

# What we should do

We should build libraries which solve the common use case and make assumptions about the input types they will be given.

## Example: scule (1.8M downloads/week)

[scule](https://www.npmjs.com/package/scule) is a library for transforming casing of text (e.g. camel case, etc).

It only accepts inputs it is designed for (strings and arrays of strings) and has zero dependencies.

In most of the functions it exports, it assumes valid input data types.

## Example: dlv (14.9M downloads/week)

[dlv](https://www.npmjs.com/package/dlv) is a library for deep property access.

It only accepts strings and arrays of strings as the path to access, and assumes this (i.e. does no validation).

# Validation is important

Validation is important, and I want to be clear that I'm not saying we should stop validating our data.

However, we should usually be validating the data in the project that owns it (e.g. at the app level), and not in every library that later consumes it as input.

> Deep dependencies applying validation like this actually shifts the burden from where it belongs (at data boundaries) to deep in the dependency tree.

Reviewer: ```suggestion
Deep dependencies applying validation like this actually shift the burden from where it belongs (at data boundaries) to deep in the dependency tree.
```

- Any are replaceable by native functionality these days (e.g. `Array.isArray`)
- Any are replaceable by smaller, more performant alternatives (e.g. `scule` instead of `pascalcase`)
- Any are redundant if you make more assumptions about input types

Tools like [npmgraph](https://npmgraph.js.org/) can help you visualise your dependency tree to make this task easier.

Also, being stricter around input types will allow you to reduce a lot of code and dependencies.

If you can assume the data being passed in is the correct type, you can leave validation up to the consumer.

## As a user

Keep a close eye on your dependencies (both deep and direct), and what alternatives are available to your direct dependencies.

Often, it is easy to stick with a dependency from long ago and forget to re-visit it one day in case there is a better way. Many of these packages are possible natively, or have more modern alternatives.

Useful tools:

- [npmgraph](https://npmgraph.js.org/) for visualising your dependency tree
- [node-modules.dev](https://node-modules.dev/) for visualising your dependencies and lots of useful meta data

> - [dependabot](https://docs.github.com/en/code-security/getting-started/dependabot-quickstart-guide) for keeping your dependencies up to date

Reviewer: ```suggestion
- [Dependabot](https://docs.github.com/en/code-security/getting-started/dependabot-quickstart-guide) for keeping your dependencies up to date
```

they use it as capital case in the docs
